File: Java/프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.java
import java.util.*;


class Solution {
    public int findCycle(int y, int x, int dir, String[] grid, int[][][] DP) {
        // 우 하 좌 상
        int[][] directions = new int[][]{{0, 1}, {-1, 0}, {0, -1}, {1, 0}};

        int sy = y, sx = x, sDir = dir;
        int yMax = grid.length, xMax = grid[0].length();
        int cycle = 1;

        while (true) {
            // 이동
            DP[y][x][dir] = 1;
            y = (yMax + y + directions[dir][0]) % yMax;
            x = (xMax + x + directions[dir][1]) % xMax;

            if (grid[y].charAt(x) == 'L') {
                dir = (4 + dir - 1) % 4;
            } else if (grid[y].charAt(x) == 'R') {
                dir = (4 + dir + 1) % 4;
            }

            if (x == sx && y == sy && sDir == dir) {
                break;
            }
            cycle += 1;
        }

        return cycle;
    }

    public int[] solution(String[] grid) {
        ArrayList<Integer> answer = new ArrayList<>();

        // DP[y][x][dir] : gird[y][x]에 dir 방향으로 빛이 통과 여부
        int[][][] DP = new int[grid.length][grid[0].length()][4];

        for (int y = 0; y < grid.length; y++) {
            for (int x = 0; x < grid[y].length(); x++) {
                for (int dir = 0; dir < 4; dir++ ) {
                    // grid[y][x] 에 dir으로 빛을 쏠 에정

                    // 이미 해당 경로가 다른 사이클에 포함되어 있음.
                    if (DP[y][x][dir] != 0) {
                        continue;
                    }

                    answer.add(findCycle(y, x, dir, grid, DP));
                }
            }
        }

        return answer.stream()
                .sorted()
                .mapToInt(Integer::intValue)
                .toArray();
    }
}

File: Python/프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.py
def find_cycle(y, x, direction, grid):
    # direction = (y, x), order = 우 하 좌 상
    directions = [(0, 1), (-1, 0), (0, -1), (1, 0)]

    s_y, s_x, s_direction = y, x, direction
    y_max, x_max = len(grid), len(grid[0])

    cycle = 1

    while True:
        # move
        grid[y][x][direction + 1] = 1
        y = (y + directions[direction][0]) % y_max
        x = (x + directions[direction][1]) % x_max
        if grid[y][x][0] == "L":
            direction = (4 + direction - 1) % 4
        elif grid[y][x][0] == "R":
            direction = (4 + direction + 1) % 4

        if y == s_y and x == s_x and s_direction == direction:
            break
        cycle += 1

    return cycle


def solution(grid):
    # (y, x)에서 상하좌우으로 빛을 쏘는 것이 다른 사이클에 포함 여부 저장
    grid = [[[grid[i][j], 0, 0, 0, 0] for j in range(len(grid[i]))] for i in range(len(grid))]

    answer = []

    for y in range(len(grid)):
        for x in range(len(grid[y])):
            for direction in range(4):

                # 해당 방향으로 빛을 쏘는 것이 다른 사이클에 포함 됨 - 중복 사이클 의미
                if grid[y][x][direction + 1] == 1:
                    continue

                # 중복 아님 -> 사이클 찾기
                answer.append(find_cycle(y, x, direction, grid))

    return sorted(answer)